### PR TITLE
add support for retrying pending changes

### PR DESCRIPTION
**Pull request by wei-lee**
# What

This PR introduces a new feature to allow the pending changes to be retried. The number of retries and the retry interval can be configured by developers.

# Why

This is related to issue: https://issues.jboss.org/browse/RHMAP-21014.

What happens is that sometimes the custom sync handlers could fail occasionally. At the moment, when an error occurred, there is no retry mechanism and the pending change is simply treated as processed and later on the client data will be reverted. This will cause some confusion to both users and developers and the sync server itself should allow occasional failures from the sync handlers.

# To test

* Run the new integration test, or
* Deploy a new cloud app and use this version of fh-sync. Then make sure the update custom handler will fail on the first call but should succeed on the subsequent calls. Run the sync app as normal, but you should see the data will be updated eventually in the backend and the client data is not reverted back.

ping @wtrocki @camilamacedo86 @shannon

### CODE REVIEW

File: lib/pending-processor.js [hunk 1]
```diff
@@ -14,6 +17,13 @@ var SYNC_UPDATE_TYPES = {
 };
 
 function saveUpdate(datasetId, pendingChange, type, msg, callback) {
+  if (type !== SYNC_UPDATE_TYPES.APPLIED && pendingChange.tries < retryLimit) {
+    //if the pending change can not be applied, but it hasn't reach the retry limit, 
+    //let's not do anything and just leave it in the queue by returning an error. 
+    //it will be tried again.
+    //evetually when pendingChange.tries == retryLimit, the right status will be saved.
+    return callback(msg || "retry");
```
Comment (wtrocki) [inline lib/pending-processor.js]: Could it be `callback(new Error(msg || "retry"));` to use the same type as in other callbacks?

Comment (wei-lee) [reply lib/pending-processor.js]: yes

File: lib/pending-processor.js [hunk 2]
```diff
@@ -250,11 +270,14 @@ function applyPendingChange(pendingChange, tries, callback) {
...
   syncStorage = syncStorageImpl;
   dataHandlers = dataHandlersImpl;
   hashProvider = hashProviderImpl;
   metricsClient = metricsClientImpl;
+  if (_.isNumber(pendingWorkerRetryLimit) && pendingWorkerRetryLimit > 0) {
+    retryLimit = pendingWorkerRetryLimit
+  }
   return function(pendingChangeRequest, callback) {
     var pendingChange = pendingChangeRequest.payload;
     var tries = pendingChangeRequest.tries;
```
Comment (wtrocki) [inline lib/pending-processor.js:283]: This works fine but initially tries will be `undefined`. Then checks for update and collision work fine because
`undefined < retryLimit` => false


Adding 0 explicitly instead of undefined actually causing bug:
```
pendingChangeRequest.tries = pendingChangeRequest.tries || 0;
var tries = pendingChangeRequest.tries`
```
When tries is 0 
`0 < retryLimit` => true and this means that conflicts that come from successful apply will not be saved.

TL;DR this code works , but we could set initial tries explicitly. 
Should we initialize tries or leave it as it is?

**Comment by wtrocki** [comment]
@wei-lee World class work. I have tested that with integration tests (with modified parameters as well) and also in standalone server.

**Review by wtrocki — APPROVED** [review]

File: lib/pending-processor.js (hunk 2 above)
Comment (wei-lee) [reply lib/pending-processor.js:283]: @wtrocki where did you see the initial value to be undefined? The mongodb queue should always set the tries value and the initial value should be 1. The initial value is set by the mongodb-queue.

Comment (wtrocki) [reply lib/pending-processor.js:283]: @wei-lee 100 % right. I did not know that this value is being set by mongo and did not checked that before making the change so we are ok

**Comment by wtrocki** [comment]
Good to go! Going to include this changes in the batch of sync changes that were done recently and retest that again.
